src/Tile/Syntax/Lifted.hs

```diff
@@ -0,0 +1,166 @@
+{-# LANGUAGE ConstraintKinds #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE TypeOperators #-}
+module Tile.Syntax.Lifted
+( Permutable
+, S.Syntax
+, var
+, wvar
+  -- * Let
+, S.Let
+, let'
+, let''
+  -- * Lam
+, S.Lam
+, lam
+, ($$)
+  -- * Type
+, S.Type
+, type'
+, (>->)
+, (-->)
+, (==>)
+  -- * Prob
+, S.Prob
+, ex
+, (===)
+  -- * Modules
+, S.Def
+  -- * Elaborator scripts
+, runScript
+, runCScript
+, throw
+, reset
+, resetC
+, liftScript
+, liftCScript
+, Script(..)
+  -- * Re-exports
+, (:::)(..)
+, Plicit(..)
+, plicit
+) where
+
+import           Control.Applicative (liftA2)
+import           Data.Distributive
+import           Tile.Functor.Compose
+import           Tile.Syntax ((:::)(..), Plicit(..), plicit)
+import qualified Tile.Syntax as S
+
+type Permutable f = (Applicative f, Distributive f)
+
+var :: Applicative m => i expr -> (m :.: i) expr
+var = C . pure
+
+wvar :: forall m i j expr . (Applicative m, Extends (m :.: i) (m :.: j)) => i expr -> (m :.: j) expr
+wvar = weakens . var @m
+
+
+-- Let
+
+let'
+  :: (Applicative m, S.Let expr, Permutable i)
+  => (m :.: i) expr ::: (m :.: i) expr
+  -> (forall j . Permutable j => (i :.: j) expr -> (m :.: i :.: j) expr)
+  -> (m :.: i) expr
+let' v f = let'' v (f . C . pure)
+
+let''
+  :: (Applicative m, S.Let expr, Permutable i)
+  => (m :.: i) expr ::: (m :.: i) expr
+  -> (forall j . Permutable j => j expr -> (m :.: i :.: j) expr)
+  -> (m :.: i) expr
+let'' (tm ::: ty) f = S.let' <$> ((:::) <$> tm <*> ty) <*> mapC (fmap getC) (f id)
+
+
+-- Lam
+
+lam :: (Applicative m, S.Lam expr, Permutable i) => (m :.: i) Plicit -> (forall j . Permutable j => (i :.: j) expr -> (m :.: i :.: j) expr) -> (m :.: i) expr
+lam p f = S.lam <$> p <*> mapC (fmap getC) (f (C (pure id)))
```

Comment:
The trick here is neat:

1. The function uses `-XRankNTypes` to ensure that the variable doesn’t escape its scope.

2. Effects occur in `m` and are limited to this specific stage.

3. We return in `m :.: i` instead of just `m` so that it’s easy to compose these things together through multiple binders.

4. `i` and `j` can be thought of as (rank-n quantified) environments; `i` is the initial environment, `j` represents the environment containing just the newly-bound variable, and `i :.: j` is their concatenation. So we’re building up successively larger environments as we descend inwards through binders.

5. We instantiate `j` to `(->) expr`, yielding (in the limit) something like `(expr -> expr) -> m (expr -> expr)`, which we apply to `id` to get an `m (expr -> expr)` which we can `<$>`/`<*>` `S.lam :: Plicit -> (expr -> expr) -> expr` over to get the final `m expr` (all of this modulo the extra `i`, of course).

6. The initial environment can be determined however you like, but is typically `Identity`. You could also use this to model the availability of globals/other non-lexically-scoped symbols.

I’m gonna explore using this stuff to model the environment at the value level too, keeping a map of bindings in e.g. the parser using whatever other information we need to pass around.